src/drivers/intel/dw-dma.c

```diff
@@ -1069,15 +1069,14 @@ static uint64_t dw_dma_work(void *data, uint64_t delay)
 
 	tracev_dma("wrk");
 
-	/* skip if channel is not running */
 	if (p->chan[i].status != COMP_STATE_ACTIVE) {
 		trace_dma_error("eDs");
-		goto out;
+		/* skip if channel is not running */
+		return 0;
 	}
 
 	dw_dma_process_block(&p->chan[i], &next);
 
-out:
 	return next.size == DMA_RELOAD_END ? 0 : p->chan[i].timer_delay;
```

Comment:
@libinyang You are just removing the error case while still keeping the trace_dma_error. That doesn't seem to address the root cause, does it?

Reply:
> @libinyang You are just removing the error case while still keeping the trace_dma_error. That doesn't seem to address the root cause, does it?

In the original code, the if clause will not set up the 'next' and goto 'out', where 'next' is used. This will cause using uninitialized value next.size. 

This patch will return 0 directly if the status is not active (We return 0 here because the status is not active so we should not reschedule this work).

With this fix, 'next' will always be set up when using the 'next.size' as 'dw_dma_process_block(&p->chan[i], &next);' is always called before using 'next.size'

Reply:
My point was: why is the function called if the status is not active? is this desired or is this a bug? I am trying to get down to the root cause, not trying to make the local issue go away.

Reply:
ok, this make sense to me.

Reply:
> My point was: why is the function called if the status is not active? is this desired or is this a bug? I am trying to get down to the root cause, not trying to make the local issue go away.

Yes, I think this function should not be called if the status is not active. We need this check in case something is wrong in somewhere else and call this function improperly. So we should stop the dma work and not reschedule it. We do need to root cause such issue if we meet.

Reply:
@libinyang  @plbossart work can only be cancelled if it's not already running. It is possible a race between stopping the DMA and scheduling the work could result in attempting to stop the work when it's running (i.e. DMA stop request immediately after a  DMA period completion).